content/posts/2025-01-15-perf-improvements.md

```diff
@@ -0,0 +1,235 @@
+
+By removing the per-CPU map, we unlock this next improvement: shrinking the map key to a lighter [7-tuple](https://github.com/netobserv/netobserv-ebpf-agent/blob/2c96c420fc5ed223b7b99b00b2705fe84c5c0110/bpf/types.h#L159-L167), performing the de-duplication directly in-kernel, resulting in reducing the map size, the CPU cost to read it, and getting rid of the whole user-space mechanism for de-duplication. Why 7-tuples, by the way? Well, it's only for some ICMP related stuff - not something to worry about, performance-wise, as ICMP traffic is generally negligible (unless your cluster is full of `pings` running everywhere!) and has a smaller flow cardinality (no ports). The most important part was to remove the interface index from the key.
+
+One of the consequences is that we would lose information about each traversed interface. We want to still be able to tell: this sample flow went through `genev_sys_6081`, then `8aa5bd1d532fca8@if2`, and finally `ens5`. So we had to add this information as an array in the map values.
+
+Less flows to fetch thanks to smaller keys, and no user-space deduplication: less used CPU.
+
+**Fig. 1 (again): eBPF agent user-space CPU cores usage, averaged per node**
+
+![User-space CPU]({page.image('perf-improvements-1-8/user-space-cpu.png')})
+
+### Other improvements
+
+There have been other improvements identified, more as low-hanging fruits, but still impactful:
+
+- Removing some redundant calls to `bpf_map_update()`
+- Removing `packed` attribute from the eBPF structures
+
+## Side effects
+
+Those changes triggered a refactoring that doesn't come without consequences and tradeoffs.
+
+- Most importantly, **partial flows**: because now having two different maps, one for the main flows, generated from the TC hook, and another for the enrichments, generated from other hooks, there can sometimes be a mismatch between these two. Especially when the enrichment map keys aren't found in the flows map, the result is a generated partial flow, which is a flow that lacks some information. Namely, the TCP flags, the MAC addresses and the bytes and packets counters are missing in partial flows. It doesn't mean these values are entirely lost; you could still be able to find them in an adjacent flow - it's because flows are evicted periodically, and an eviction might occur precisely at the _wrong moment_ (it's a race condition), with only partial data being available at that time. Another cause for partial flows could be sampling, when the agent is configured with a sampling rate: if for some reason the enrichment data is sampled but the corresponding TC hook packet isn't, this would also result in a partial flow.
```

Comment:
Reword sentence to:
Especially when the enrichment map keys aren't found in the flows map, the result is a generated partial flow, which is a flow that lacks some information, namely the TCP flags, the MAC addresses, and the bytes and packets counters.

Reword sentences to:
Another cause for partial flows is when the agent is configured with a sampling rate greater than one.  If, for some reason, the enrichment data is sampled but the corresponding TC hook packet isn't, this would also result in a partial flow.